test/utils.py

```diff
@@ -633,3 +705,208 @@ def relative_l_inf_error(a_true: ndarray, a: ndarray) -> float:
             converged = rel_error < target_rel_error
 
     assert converged
+
+
+def _test_inplace_activations(
+    linop_cls: Type[Union[KFACLinearOperator, EKFACLinearOperator]], dev: device
+):
+    """Test that (E)KFAC works if the network has in-place activations.
+
+    We use a test case with a single datum as (E)KFAC becomes exact as the number of
+    MC samples increases.
+
+    Args:
+        linop_cls: The linear operator class to test.
+        dev: The device to run the test on.
+    """
+    manual_seed(0)
+    model = Sequential(Linear(6, 3), ReLU(inplace=True), Linear(3, 2)).to(dev)
+    loss_func = MSELoss().to(dev)
+    batch_size = 1
+    data = [(rand(batch_size, 6), regression_targets((batch_size, 2)))]
+    params = list(model.parameters())
+
+    # 1) compare (E)KFAC and GGN
+    ggn = block_diagonal(
+        GGNLinearOperator, model, loss_func, params, data, return_numpy=False
+    )
+
+    linop = linop_cls(model, loss_func, params, data, mc_samples=2_000)
+    linop_mat = linop @ eye(linop.shape[1])
+
+    atol = {"sum": 5e-1, "mean": 2e-3}[loss_func.reduction]
+    rtol = {"sum": 2e-2, "mean": 2e-2}[loss_func.reduction]
+
+    assert allclose_report(ggn, linop_mat, rtol=rtol, atol=atol)
+
+    # 2) Compare GGN (inplace=True) and GGN (inplace=False)
+    for mod in model.modules():
+        if hasattr(mod, "inplace"):
+            mod.inplace = False
+    ggn_no_inplace = block_diagonal(
+        GGNLinearOperator, model, loss_func, params, data, return_numpy=False
+    )
+
+    assert allclose_report(ggn, ggn_no_inplace)
+
+
+def _test_property(  # noqa: C901
+    linop_cls: Type[Union[KFACLinearOperator, EKFACLinearOperator]],
+    property_name: str,
+    model: Module,
+    loss_func: Module,
+    params: List[Parameter],
+    data: Iterable[Tuple[Union[Tensor, MutableMapping], Tensor]],
+    batch_size_fn: Optional[Callable[[MutableMapping], int]],
+    separate_weight_and_bias: bool,
+    check_deterministic: bool,
+    rtol: float = 1e-5,
+    atol: float = 1e-8,
+):
```

Comment:
This currently does not test different KFAC approximations and Fisher types. Is that okay?

Reply:
I think so since how these properties are computed only depends on the Kronecker factors and not on how they were computed.